File: gameserver/network/packet.py
from abc import abstractmethod, ABC
from typing import Callable, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from gameserver.client import GameClient
    from gameserver.network.utils.reader import Reader


class Packet(ABC):
    PACKET_ID = 0

    def __init__(self):
        self.data = bytearray()
        self.packet_id = self.PACKET_ID
        self.packet_size = 0
        self.reader: Optional[Reader] = None

    @abstractmethod
    def finalize(self):
        pass

    def to_hex(self):
        return self.reader.to_hex_string()

    def parse_packet(self, ):
        raise NotImplementedError

    @staticmethod
    def parse_packet_data(packet_size, reader, packet: 'Callable[[],Packet]'):
        p = packet()
        p.reader = reader
        p.buffer = reader.buffer
        p.packet_size = packet_size
        p.parse_packet()
        return p

    def handle_packet(self, client: 'GameClient'):
        raise NotImplementedError

File: src/network/packet.js
import Reader from './utils/reader.js';
import Writer from './utils/writer.js';

class Packet {
    constructor() {
        this.data = null;
        this.packetId = -1;
        this.packetSize = 0;
        this.reader = null
    }


    setPacketData(data) {
        this.data = data;
    }


    toHexString() {
        if (this.reader === null)
            throw new Error("Reader is null");

        return this.reader.toHexString();
    }


    parsePacket() {
        throw new Error("Not implemented");
    }

    static parsePacketData(packetSize, reader, packet) {
        let p = new packet();
        p.reader = reader;
        p.data = reader.data;
        p.packetSize = packetSize;
        p.parsePacket();
        return p;
    }

    handleReceivedPacket(client) {
        throw new Error("Not implemented");
    }

    finalize() {
        throw new Error("Not implemented");
    }

}

export default Packet;